libs/vgc/widgets/console.cpp

```diff
@@ -453,6 +455,7 @@ void Console::keyPressEvent(QKeyEvent* e)
 
         // This has to be called after any navigation of the cursor
```

Comment:
Are these three lines still needed at all now that you added setReadOnly(false)?

Reply:
Hum... this seems strange to me. These three lines of code expand to either `setReadOnly(true); setReadOnly(false);` or `setReadOnly(false); setReadOnly(false);` depending on the position of the cursor/selection. The first two lines of code in `protectPreviousCodeBlocks_` shouldn't have any side-effects on the text edit since they only create and modify the local variable `cursor`.

Reply:
Yeah, it really does not make any sense if you just look at the code. But I just tested all possible combinations. It appears that both are needed to have the desired result.

Without both:
The first time I write something inside previous block, the dialog with chinese characters appears. But nothing gets entered. Every other time the dialog does not show up. However the first time I write inside last block, the first character entered is written as normal english character, then the next are using the chinese dialog. That would be a bug.

With only setReadOnly(false):
The first time I write something inside previous block, is like above. But the bug is now fixed.

With only protectPreviousCodeBlocks_():
Now the cursor is not display inside previous block if I move with arrow keys. If I type something it shows up again.

With both:
Nothing appears inside previous block and cursor is always visible. And in last block write is still possible.

Reply:
Thanks for the explanation, I think we'll have to refactor a bit. I'm currently writing a numbered list of desired behaviours, so that we can exhaustively test each change and discuss the bugs more effectively. I'll get back to this problem after I finished writing this.